File: chain/java/ChainOfResponsibility.java
import java.util.ArrayList;
import java.util.List;

class Content {
    private String content;

    public Content(String content) {
        this.content = content;
    }

    public String get() {
        return content;
    }

    public void set(String content) {
        this.content = content;
    }
}

// ============================================================

interface SensitiveWordFilter1 {
    boolean doFilter(Content content);
}

class SexyWordFilter1 implements SensitiveWordFilter1 {
    @Override
    public boolean doFilter(Content content) {
        System.out.println("1. 「情色内容」过滤中。。。");
        boolean legal = !content.get().contains("涉黄");
        legal = legal && !content.get().contains("儿童不宜");
        // 返回内容是否合法，true 则是合法的，会继续执行链中的下一个过滤器（如果有的话）
        return legal;
    }
}

class PoliticalWordFilter1 implements SensitiveWordFilter1 {
    @Override
    public boolean doFilter(Content content) {
        System.out.println("1. 「政治敏感内容」过滤中。。。");
        boolean legal = !content.get().contains("反动");
        legal = legal && !content.get().contains("不符合社会价值观");
        return legal;
    }
}

class AdsWordFilter1 implements SensitiveWordFilter1 {
    @Override
    public boolean doFilter(Content content) {
        System.out.println("1. 「广告内容」过滤中。。。");
        return !content.get().contains("广告");
    }
}

class SensitiveWordFilterChain1 {
    // 使用集合存储所有过滤器
    private final List<SensitiveWordFilter1> filters = new ArrayList<>();

    public void addFilter(SensitiveWordFilter1 filter) {
        filters.add(filter);
    }

    /**
     * 过滤内容
     *
     * @return true 如果内容不包含敏感词
     */
    public boolean filter(Content content) {
        for (SensitiveWordFilter1 filter : filters) {
            if (!filter.doFilter(content)) {
                return false;
            }
        }
        return true;
    }
}

// ============================================================

interface SensitiveWordFilter2 {
    void doFilter(Content content);
}

abstract class Successor implements SensitiveWordFilter2 {
    protected Successor successor;  // next 后缀过滤器

    public void setSuccessor(Successor successor) {
        this.successor = successor;
    }

    public final void filter(Content content) {
        doFilter(content);
        if (successor != null) {
            // 另一种变体，所有请求都会被所有过滤器过滤一遍，不存在中途终止的情况
            successor.filter(content);
        }
    }
}

class SexyWordFilter2 extends Successor {
    @Override
    public void doFilter(Content content) {
        System.out.println("2. 「情色内容」过滤中。。。");
        String c = content.get().replace("涉黄", "**").replace("儿童不宜", "****");
        content.set(c);
    }
}

class PoliticalWordFilter2 extends Successor {
    @Override
    public void doFilter(Content content) {
        System.out.println("2. 「政治敏感内容」过滤中。。。");
        String c = content.get().replace("反动", "##").replace("不符合社会价值观", "########");
        content.set(c);
    }
}

class AdsWordFilter2 extends Successor {
    @Override
    public void doFilter(Content content) {
        System.out.println("2. 「广告内容」过滤中。。。");
        String c = content.get().replace("广告", "$$");
        content.set(c);
    }
}

class SensitiveWordFilterChain2 {
    // 使用链表存储过滤器，head 表示当前执行到的过滤器，tail 表示过滤器链的末尾过滤器（方便在链表末尾添加新的过滤器）
    private Successor head;

    private Successor tail;

    public void addFilter(Successor filter) {
        filter.setSuccessor(null);  // 保证它不会自带后继过滤器（否则 tail 就需要递归获取到最后的过滤器来重新指定）
        // 初次添加首个过滤器的情况
        if (head == null) {
            head = filter;
            tail = filter;
            return;
        }
        tail.setSuccessor(filter);
        tail = filter;
    }

    public void filter(Content content) {
        if (head != null) {
            head.filter(content);
        }
    }
}

// ============================================================

interface SensitiveWordFilter3 {
    void doFilter(Content content, SensitiveWordFilterChain3 chain);
}

class SexyWordFilter3 implements SensitiveWordFilter3 {
    @Override
    public void doFilter(Content content, SensitiveWordFilterChain3 chain) {
        System.out.println("3.1 「情色内容」过滤中。。。");
        String c = content.get().replace("涉黄", "**").replace("儿童不宜", "****");
        content.set(c);
        System.out.println("3.1 拦截请求，请求前的帖子内容为：" + content.get());
        chain.filter(content);
        System.out.println("3.1 拦截响应，响应前的帖子内容为：" + content.get());
    }
}

class PoliticalWordFilter3 implements SensitiveWordFilter3 {
    @Override
    public void doFilter(Content content, SensitiveWordFilterChain3 chain) {
        System.out.println("3.2 「政治敏感内容」过滤中。。。");
        String c = content.get().replace("反动", "##").replace("不符合社会价值观", "########");
        content.set(c);
        System.out.println("3.2 拦截请求，请求前的帖子内容为：" + content.get());
        chain.filter(content);
        System.out.println("3.2 拦截响应，响应前的帖子内容为：" + content.get());
    }
}

class AdsWordFilter3 implements SensitiveWordFilter3 {
    @Override
    public void doFilter(Content content, SensitiveWordFilterChain3 chain) {
        System.out.println("3.3 「广告内容」过滤中。。。");
        String c = content.get().replace("广告", "$$");
        content.set(c);
        System.out.println("3.3 拦截请求，请求前的帖子内容为：" + content.get());
        chain.filter(content);
        System.out.println("3.3 拦截响应，响应前的帖子内容为：" + content.get());
    }
}

interface Something {
    void justDoIt(Content content);
}

class SensitiveWordFilterChain3 {
    // 使用数组存储所有过滤器
    private SensitiveWordFilter3[] filters = new SensitiveWordFilter3[]{};

    private int pos = 0; // 当前执行到了哪个 filter
    private int n;       // filter 的个数

    private final Something something;

    private static final int INCREMENT = 16;

    public SensitiveWordFilterChain3(Something something) {
        this.something = something;
    }

    public void addFilter(SensitiveWordFilter3 filter) {
        for (SensitiveWordFilter3 f : filters) {
            if (filter == f) return;
        }

        // 扩容
        if (n == filters.length) {
            SensitiveWordFilter3[] newFilters = new SensitiveWordFilter3[n + INCREMENT];
            System.arraycopy(filters, 0, newFilters, 0, n);
            filters = newFilters;
        }

        filters[n++] = filter;
    }

    public void filter(Content content) {
        if (pos < n) {
            SensitiveWordFilter3 filter = filters[pos++];
            // 这里其实是递归调用 chain.filter，参考 Tomcat 的过滤器链的实现和使用去理解
            filter.doFilter(content, this);
        } else {
            // filter 都处理完毕后，执行业务
            something.justDoIt(content);
        }
    }
}

File: chain/go/chain_of_responsibility.go
package chain

import (
	"fmt"
	"strings"
)

type Content struct {
	Content string
}

// ============================================================

type SensitiveWordFilter1 interface {
	doFilter(content Content) bool
}

type SexyWordFilter1 struct{}

func (f SexyWordFilter1) doFilter(content Content) bool {
	fmt.Println("1. 「情色内容」过滤中。。。")
	legal := !strings.Contains(content.Content, "涉黄")
	legal = legal && !strings.Contains(content.Content, "儿童不宜")
	return legal
}

type PoliticalWordFilter1 struct{}

func (f PoliticalWordFilter1) doFilter(content Content) bool {
	fmt.Println("1. 「政治敏感内容」过滤中。。。")
	legal := !strings.Contains(content.Content, "反动")
	legal = legal && !strings.Contains(content.Content, "不符合社会价值观")
	return legal
}

type AdsWordFilter1 struct{}

func (f AdsWordFilter1) doFilter(content Content) bool {
	fmt.Println("1. 「广告内容」过滤中。。。")
	return !strings.Contains(content.Content, "广告")
}

type SensitiveWordFilterChain1 struct {
	filters []SensitiveWordFilter1
}

func (c *SensitiveWordFilterChain1) AddFilter(filter1 SensitiveWordFilter1) {
	c.filters = append(c.filters, filter1)
}

func (c *SensitiveWordFilterChain1) Filter(content Content) bool {
	for _, filter := range c.filters {
		if !filter.doFilter(content) {
			return false
		}
	}
	return true
}

// ============================================================

type SensitiveWordFilter2 interface {
	doFilter(content *Content)
}

type Successor struct {
	doFilter  func(content *Content)
	successor *Successor // next 后缀过滤器
}

func (f *Successor) SetSuccessor(successor *Successor) {
	f.successor = successor
}

func (f *Successor) Filter(content *Content) {
	f.doFilter(content)
	if f.successor != nil {
		f.successor.Filter(content)
	}
}

type SexyWordFilter2 struct{}

func (f *SexyWordFilter2) doFilter(content *Content) {
	fmt.Println("2. 「情色内容」过滤中。。。")
	c := strings.ReplaceAll(content.Content, "涉黄", "**")
	content.Content = strings.Replace(c, "儿童不宜", "****", -1) // the same as `strings.ReplaceAll(c, "儿童不宜", "****")`
}

type PoliticalWordFilter2 struct{}

func (f *PoliticalWordFilter2) doFilter(content *Content) {
	fmt.Println("2. 「政治敏感内容」过滤中。。。")
	c := strings.ReplaceAll(content.Content, "反动", "##")
	content.Content = strings.ReplaceAll(c, "不符合社会价值观", "########")
}

type AdsWordFilter2 struct{}

func (f *AdsWordFilter2) doFilter(content *Content) {
	fmt.Println("2. 「广告内容」过滤中。。。")
	content.Content = strings.ReplaceAll(content.Content, "广告", "$$")
}

type SensitiveWordFilterChain2 struct {
	head *Successor
	tail *Successor
}

func (c *SensitiveWordFilterChain2) AddFilter(filter2 SensitiveWordFilter2) {
	successor := &Successor{filter2.doFilter, nil}
	if c.head == nil {
		c.head = successor
		c.tail = successor
		return
	}
	c.tail.SetSuccessor(successor)
	c.tail = successor
}

func (c *SensitiveWordFilterChain2) Filter(content *Content) {
	if c.head != nil {
		c.head.Filter(content)
	}
}

// ============================================================

type SensitiveWordFilter3 interface {
	doFilter(content *Content, chain *SensitiveWordFilterChain3)
}

type SexyWordFilter3 struct{}

func (f *SexyWordFilter3) doFilter(content *Content, chain *SensitiveWordFilterChain3) {
	fmt.Println("3.1 「情色内容」过滤中。。。")
	c := strings.ReplaceAll(content.Content, "涉黄", "**")
	content.Content = strings.ReplaceAll(c, "儿童不宜", "****")
	fmt.Println("3.1 拦截请求，请求前的帖子内容为：", content.Content)
	chain.Filter(content)
	fmt.Println("3.1 拦截响应，响应前的帖子内容为：", content.Content)
}

type PoliticalWordFilter3 struct{}

func (f *PoliticalWordFilter3) doFilter(content *Content, chain *SensitiveWordFilterChain3) {
	fmt.Println("3.2 「政治敏感内容」过滤中。。。")
	c := strings.ReplaceAll(content.Content, "反动", "##")
	content.Content = strings.ReplaceAll(c, "不符合社会价值观", "########")
	fmt.Println("3.2 拦截请求，请求前的帖子内容为：", content.Content)
	chain.Filter(content)
	fmt.Println("3.2 拦截响应，响应前的帖子内容为：", content.Content)
}

type AdsWordFilter3 struct{}

func (f *AdsWordFilter3) doFilter(content *Content, chain *SensitiveWordFilterChain3) {
	fmt.Println("3.3 「广告内容」过滤中。。。")
	content.Content = strings.ReplaceAll(content.Content, "广告", "$$")
	fmt.Println("3.3 拦截请求，请求前的帖子内容为：", content.Content)
	chain.Filter(content)
	fmt.Println("3.3 拦截响应，响应前的帖子内容为：", content.Content)
}

type Something interface {
	JustDoIt(content *Content)
}

type SensitiveWordFilterChain3 struct {
	filters []SensitiveWordFilter3

	pos int // 当前执行到了哪个 filter
	n   int // filter 的个数

	something Something
}

func NewSensitiveWordFilterChain3(something Something) *SensitiveWordFilterChain3 {
	return &SensitiveWordFilterChain3{something: something}
}

func (c *SensitiveWordFilterChain3) AddFilter(filter3 SensitiveWordFilter3) {
	for _, f := range c.filters {
		if filter3 == f {
			return
		}
	}
	c.filters = append(c.filters, filter3)
	c.n++
}

func (c *SensitiveWordFilterChain3) Filter(content *Content) {
	if c.pos < c.n {
		filter3 := c.filters[c.pos]
		c.pos++
		filter3.doFilter(content, c)
	} else {
		c.something.JustDoIt(content)
	}
}
